clientcredentialsx/auth.go

```diff
@@ -0,0 +1,70 @@
+package clientcredentialsx
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/spf13/pflag"
+	"github.com/spf13/viper"
+
+	"go.infratographer.com/x/viperx"
+
+	"golang.org/x/oauth2/clientcredentials"
+)
+
+// AuthConfig is an oauth2 client config
+type AuthConfig struct {
+	// ClientID is the applications's ID
+	ClientID string
+
+	// ClientSecret is the application's secret
+	ClientSecret string
+
+	// TokenURL is the resource server's token endpoint URL. This is a constant specific to each server.
+	TokenURL string
+
+	// Scopes specifies optional requested permissions
+	Scopes []string
+}
+
+// Auth handles client credentials oauth2 authentication
+type Auth struct {
+	cfg clientcredentials.Config
+}
+
+// NewAuth returns a new oauth2 http client
+func NewAuth(cfg AuthConfig) *Auth {
+	ccCfg := clientcredentials.Config{
+		ClientID:     cfg.ClientID,
+		ClientSecret: cfg.ClientSecret,
+		TokenURL:     cfg.TokenURL,
+		Scopes:       cfg.Scopes,
+	}
+
+	cli := &Auth{
+		cfg: ccCfg,
+	}
+
+	return cli
+}
+
+// HTTPClient returns an http client using the configured client credentials config.
+// The token will auto-refresh as necessary.
+func (a Auth) HTTPClient(ctx context.Context) *http.Client {
+	return a.cfg.Client(ctx)
+}
+
+// MustViperFlags adds oidc oauth2 client config to the provided flagset and binds to viper
+func MustViperFlags(v *viper.Viper, flags *pflag.FlagSet) {
+	flags.Bool("oidc-client", true, "use oidc client auth")
+	viperx.MustBindFlag(v, "oidc.client.enabled", flags.Lookup("oidc-client"))
+
+	flags.String("oidc-client-id", "", "expected oidc client identifier")
+	viperx.MustBindFlag(v, "oidc.client.id", flags.Lookup("oidc-client-id"))
```

Comment:
We already use the `oidc` prefix elsewhere: https://github.com/infratographer/x/blob/67c1a1e705ac3695c31140fd9044ed301fd6c93e/echojwtx/flags.go#L22-L34

I'll have to look elsewhere in the code, but generally the convention thus far has been that each package in `x` has a corresponding top-level prefix for flags. I'm not sure what we would want to use here though...

Reply:
yah, i was supposed to remove that

Reply:
https://github.com/infratographer/x/pull/90#discussion_r1210895465

The idea is to eventually have things look like this
```
oidc:
  client:
    id: client-id-x
  server:
    enabled: true
```